### ISSUE TITLE
[FEATURE] Make aws-sdk and uuid a peer dependency

### ISSUE THREAD

**Issue by MickL** [dynamoose/dynamoose#850]
Both are frequently updated and the version should not be dependent of Dynamoose to be recently updated.

**Comment by fishcharlie** [comment]
@MickL The problem with peer dependencies is that they are not installed alongside the package. Which turns an install command of:

```
npm install dynamoose
```

into

```
npm install dynamoose aws-sdk uuid
```

The major question here is if it's valuable to allow users to set a custom version of these dependencies. I would argue in this case, no, that is not valuable. A lot of the frequent `aws-sdk` updates do not effect DynamoDB at all. And for the most part we keep our dependencies up to date pretty well.

I don't believe the question is in regards to frequency of updates and keeping dependencies up to date. If the concern is keeping dependencies up to date, then there are better solutions (see below). If the concern is allowing the user to set a custom version for `uuid` and `aws-sdk` then yes, `peerDependencies` is potentially the way to go here. I argue that the second case of letting a user set custom versions is unnecessary. So that being said, here are some ideas for keeping dependencies up to date:

I would be more than happy to move to a system of setting the version of `aws-sdk` for example to `2.x` in the `package.json` file. However, even better. Would be running a cron job that runs [`updtr`](https://www.npmjs.com/package/updtr) every night, and submits a PR if any dependencies need to be updated. Potentially even automatically merging that PR when all the checks pass. Maybe also in the event of a failure, it could create an issue that the dependencies need to be automatically updated?? Some cool potential for automation here maybe.

I believe `peerDependencies` should be reserved for plugins for example. A great example of this is soon I want to bring back plugin support to Dynamoose in v2.x. So plugins for Dynamoose could set Dynamoose as a peer dependency of `2.x` and indicate that they require that, assuming that the user already has Dynamoose preinstalled already (which if you are using a Dynamoose plugin, is a reasonable assumption).

Reference:

- https://docs.npmjs.com/files/package.json#peerdependencies

**Comment by hardyscc** [comment]
> I would be more than happy to move to a system of setting the version of aws-sdk for example to 2.x in the package.json file. However, even better. Would be running a cron job that runs updtr every night, and submits a PR if any dependencies need to be updated. Potentially even automatically merging that PR when all the checks pass. Maybe also in the event of a failure, it could create an issue that the dependencies need to be automatically updated?? Some cool potential for automation here maybe.

@fishcharlie Yes, you should setup [renovate](https://github.com/renovatebot/renovate) to automate the dependencies upgrade, and enable auto-merge when all the Github Actions success.
I did it on nestjs-dynamoose project, please see https://github.com/hardyscc/nestjs-dynamoose/pulls?q=is%3Apr+is%3Aclosed for the sample of auto-merged PR :)

**Comment by fishcharlie** [comment]
@hardyscc Part of me wishes this was built into GitHub Actions more. Looks like it is a GitHub App not an action. The huge benefit to this would be that it shows details in the PR about the dependencies that were updated. To replicate that, would require a decent amount of effort.

We went through the process of adding a bunch of complexity and 3rd party tools toward the end of v1. It made managing the project so much easier. But in some aspects it made it much worse. When something failed, we didn't have as much control over why. And it wasn't as customizable as it was before. You'll notice in the CHANGELOG.md file, there was a period where the changelog looked super different. And during that same period if you go into the releases tab on GitHub there wasn't any information.

Since then, I've become a much bigger fan of more in house tooling. It fits the project better, and we have more control over it.

**Comment by IMBurbank** [comment]
First, a million thanks @fishcharlie for all you've put in on this project. This tool is a great compliment to mongoose in our app.

Adding to the issues brought up by @MickL , using the `aws-sdk` as a minor-pinned dependency can create bloated productions containers and a risk of undefined behavior due to multiple versions of the sdk being used in the same app. 

There are several useful projects in the npm ecosystem (such as this and winston-cloudwatch) that provide very useful functionality, and are otherwise quite small - if not for `aws-sdk`. The sdk is 30-50MB per install and, when the various dependencies fall out of sync, this package duplication can represent half the size of our production app. 

I think using a peer dependency, pinning to a major version or rolling a continuous update pipeline could work. The continuous update wouldn't keep this minor version in sync with other popular sdk tools in the ecosystem, but it would make the job much easier.

**Comment by MickL** [comment]
> npm install dynamoose aws-sdk uuid

This is exactly what should be totally fine. Check for example ng-bootstrap, it will not install bootstrap for you and thats a good thing. I agree with @IMBurbank

**Comment by fishcharlie** [comment]
@MickL 

> This is exactly what should be totally fine.

I disagree 100%. Adding more complexity to install is not something I support at all. When I run `npm i dynamoose`, it should install everything you need. Dynamoose isn't a plugin for AWS-SDK or uuid.

> Check for example ng-bootstrap, it will not install bootstrap for you and thats a good thing.

`ng-bootstrap` looks to be locked to a specific Bootstrap version for a specific version. And according to the Installation section, you don't have to install Bootstrap separately. So either the documentation is poor, or it's independent of each other. Also that package has 0 peerDependencies, so that isn't the best example.

> I agree with @IMBurbank

Also, @MickL you said the following before in the Slack.

> Its a node.js module. Imo it does not matter what it imports.
> Some modules have hundrets of dependencies

I think that is completely contradictory to what @IMBurbank said. As I said in Slack `I’d prefer to keep the package size down`. So I do agree with some of the points that @IMBurbank made. But I was under the impression that this was not due to package size like @IMBurbank mentioned, but in fact due to version pinning.

I would love to hear from you @MickL what the priority is here.


---

@IMBurbank 

> using the aws-sdk as a minor-pinned dependency can create bloated productions containers and a risk of undefined behavior due to multiple versions of the sdk being used in the same app.

This should be a nonissue so long as `aws-sdk` conforms to semver, correct?? Not sure I'm following here.

---

Here is my proposal for how to fix this:

1. Change the dependency version of `aws-sdk` to be `2.x`
2. Run nightly cron job to use `updtr` and submit PRs for that using GitHub Actions

Thoughts?

**Comment by MickL** [comment]
Running `npm install dynamoose aws-sdk uuid —save` is not even an extra step. As @IMBurbank said I dont want to have multiple versions of the same library in my project. This is the main point a peer dependency tries to solve. 

I do have aws-sdk already and want to use this version for Dynamoose, too. And not to have it install and use a different version. Also I dont want to be dependent on Dynamoose what aws-sdk version I use. Maybe I want latest, maybe I need to downgrade because of a new bug.

Ng-bootstrap was just one of hundreds of examples. E.g. ngx-quill requires quill and rxjs to be installed. I dont know what you mean it is not mentioned in the docs? The [documentation](https://ng-bootstrap.github.io/#/getting-started) says:
> Add bootstrap and @ng-bootstrap/ng-bootstrap dependencies from npm

But also if you would install ng-bootstrap oder ngx-quill npm will tell you if it requires a peer dependency that is not installed.

**Comment by IMBurbank** [comment]
@fishcharlie I think that would work just fine. This will make it a lot easier to update `aws-sdk` along with community tools such as this - and come out with a single instance of the sdk. In this case, even downgrading would still result in one version of the sdk.

**Comment by fishcharlie** [comment]
Tl;dr, this works as expected and all use cases mentioned in the thread are handled properly with no additional changes required from Dynamoose. Closing this issue.

---

Ok so I just ran some tests here to investigate this further. I had never run into this, and I was pretty sure NPM handles this correctly and is smart enough to handle this, but I didn't have the exact details or specific tests run to back it up, which is why I held off saying this until I got time to test it.

Because dynamoose dependencies are `^` it will allow for the latest aws-sdk updates without requiring dynamoose to update our dependencies (please read link in references section for more information about `^`). For example, if I have the following dependencies:

```
"aws-sdk": "2.667.0",
"dynamoose": "^2.1.0"
```

It will install `aws-sdk` 2.667.0, and Dynamoose will use that same version since dynamoose is set to version `"aws-sdk": "^2.665.0"`. It will **NOT** install 2 versions of `aws-sdk`.

The **only** way you run into duplicate versions of `aws-sdk` is:

1. If you have a pinned version of `aws-sdk` less than the `dynamoose` `aws-sdk` dependencies version. (ex. dynamoose is `"aws-sdk": "^2.665.0"` and yours is `"aws-sdk": "2.660.0"`). In this case it will install 2 copies of `aws-sdk` `2.668.0` (as of writing this) AND `2.660.0` One way to solve this is to ensure your version also uses the `^` (if yours is `"aws-sdk": "2.660.0"` and Dynamoose is `"aws-sdk": "^2.665.0"` then as of writing this it will install one version: 2.668.0, no duplicates).
2. If `aws-sdk` bumps to a major version (3.x.x) and Dynamoose is using 2.x.x and you are using 3.x.x or you are using 2.x.x and Dynamoose is using 3.x.x. This is highly unlikely tho. `aws-sdk` is working on v3, but it looks to be housed in separate packages. Totally separate conversation tho and would be a breaking change to Dynamoose. So much bigger deal anyways. Non issue and irrelevant to the concerns brought up in this thread as far as I'm concerned.

In short, it's **extremely** rare that you will have duplicate versions of dependencies installed, unless you are pinning versions. Which in that case, it's your responsibility to keep your dependencies up to date. Dynamoose dependency versions are extremely flexible, and cover a lot of versions.

On some level you would prefer Dynamoose to update our dependencies less often, since if you have pinned versions, that will increase the chance of duplicate versions.

Also in the case of `aws-sdk` `2.x`. That is the same as `^2.667.0` in terms of future versions, UNLESS you have a pinned version and are not using the `^`.

---

So after this test. I've changed my mind.

1. I will not be changing the `aws-sdk` version to `2.x` in Dynamoose. This is to ensure that if we use a newer feature of `aws-sdk` and you are using an extremely old version of `aws-sdk` that there isn't incompatibility issues. For example, when transactions support came out, that required a new version of `aws-sdk`. Saying that Dynamoose is compatible with all `2.x` versions of `aws-sdk` is just false, and not accurate, since we are taking advantage of newer aws-sdk features.
2. I will be closing this issue as it is not actionable.
3. In the future I will consider automating updating of dependencies. We are pretty good about updating dependencies right now, but would increase the speed to update dependencies. But as I said above, on some level if you are pinning versions, you don't want us to update dependencies as frequently.
4. If you still have issues with duplicate dependency installs, please create a minimal reproducible example, and comment here. I will do my best to help try to pinpoint why you are getting duplicate dependencies installed. The best way to do this is:
    1. Create a new empty directory
    2. `npm init -y`
    3. Modify your `package.json` to include **only** the relevant packages in your `dependencies` to reproduce the example (ex. dynamoose, aws-sdk, uuid, etc). It should only include `dynamoose` and dependencies of `dynamoose`.
    4. Run `npm install`
    5. Ensure the folder structure of `node_modules` is accurate and includes duplicate dependencies. If it does not, try deleting `node_modules` and `package-lock.json` and rerun step 4.
    6. Zip up the folder, and comment here with it.

In short. I'm confident this is working correctly as is. The major concerns I heard were:

1. Not wanting duplicate versions of `aws-sdk` (in the event you have `dynamoose` & `aws-sdk` in your project dependencies)
2. Risk of undefined behavior due to multiple versions of the sdk being used in the same app
    - Didn't really address this one above. But Node.js is smart enough to use the dependency defined. In some other package managers incompatible versions would throw an error message. NPM/Node is smart enough to install multiple versions in this case. I can't think of an example of how this would even be possible with how Node/NPM works. Also side note, `aws-sdk` followers semver extremely well, so that is also a non issue as well.
3. Risk of Dynamoose not using latest `aws-sdk` version since it's updated frequently

All of these concerns are non issues, and the current behavior will not lead to issues in any one of these items. All of my reasoning for this is above.

The only concern that I heard that has some merit behind it, is if a new version of `aws-sdk` includes a new bug, and you have to pin your version to an older version of `aws-sdk`. Here is why I believe that is a non-issue tho.

1. I have personally never experienced a new bug being released into a minor version of `aws-sdk`. In the event that it does, it gets patched so quickly that I don't even notice it. This is extremely rare.
2. In the event that does happen and it's a blocker for you, you can always pin your version of `aws-sdk` to an older version. This does have the potential to install two copies of `aws-sdk`, but since a patch is likely to be released quickly, it's only temporary.
3. When updating dependencies **ALL** Dynamoose tests are run. Currently there isn't many full end to end tests of Dynamoose working with `aws-sdk`. This is something that needs to be improved, and something that is on my todo list to work on. This would also increase the chances of catching these bugs earlier on and preventing updating `aws-sdk`. Of course, PRs are welcome here.

In short, this one specific use case about bugs being introduced to `aws-sdk` is rare, uncommon, and easily fixable. I don't think it's reasonable to change the behavior of `dynamoose` to handle this one extremely rare use case. The real solution here if we want to improve this use case is to get more end to end tests of `dynamoose` integrated so we ensure that when we update the dependencies that everything is working correctly (would love some help on this!).

Please let me know if you have any additional concerns of questions. But I'm confident after testing this and playing around with it further that all of the concerns brought up in this thread are non-issues.

---

References:

- https://docs.npmjs.com/misc/semver#caret-ranges-123-025-004

**Comment by MickL** [comment]
Or you could just make it a peer-dependency so I can decide whatever version I want to use. No need to hope npm is grabbing the same versions for both projects. It could be so simple...

**Comment by fishcharlie** [comment]
@MickL 

> Or you could just make it a peer-dependency so I can decide whatever version I want to use

Please **read** what I said above. I gave direct and very specific reasons for why using a peerDependency is not correct in this use case and would create problems. I'd suggest you review how peerDependencies and dependencies in general work in NPM. Peer dependencies would create problems in this scenario.

> No need to hope npm is grabbing the same versions for both projects

There is no hope here. This is technology. Computers do exactly what they are programmed to do. In the event this isn't working, that would be a bug you should file with NPM. Not Dynamoose. But I've **never** run into that problem before. NPM is very stable and very smart about how it handles dependencies.

> It could be so simple

It is 😃. `npm install dynamoose`. Extremely simple straight forward, and gets you everything you need.

---

Please also read the section under:

> If you still have issues with duplicate dependency installs, please create a minimal reproducible example, and comment here. I will do my best to help try to pinpoint why you are getting duplicate dependencies installed.

Please follow that guideline or I will have to lock this issue. I'm unwilling to debate this any further unless you reply directly to the points I made in https://github.com/dynamoose/dynamoose/issues/850#issuecomment-623050097. I spent a lot of time and effort writing that message and verifying that everything I said was correct. I'm totally willing to engage and hear the community out on this, but you have to reply to what I said and give factual reasons and responses to my comment. Right now, I'm 100% sure we are handling this correctly. If you want me to reassess this, please take the time to reply to my comment and give facts and proof for why we should change this behavior.
